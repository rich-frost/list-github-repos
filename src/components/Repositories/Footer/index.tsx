import { Tr, Td, Tfoot, Text, Flex } from '@chakra-ui/react';

export default function Footer() {
  return (
    <Tfoot>
      <Tr>
        <Td colSpan={3}>
          <Flex justify="flex-end">
            <Text as="i">
              Showing top 10 GitHub repos for topic, sorted by best match
            </Text>
          </Flex>
        </Td>
      </Tr>
    </Tfoot>
  );
}
