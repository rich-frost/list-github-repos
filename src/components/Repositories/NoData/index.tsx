import { Tr, Td, Icon, Center } from '@chakra-ui/react';
import { BiError } from 'react-icons/bi';

export default function NoData() {
  return (
    <Tr>
      <Td colSpan={3}>
        <Center>
          <Icon as={BiError} mx="5px" /> No repositories found
        </Center>
      </Td>
    </Tr>
  );
}
