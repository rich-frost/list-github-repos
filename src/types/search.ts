export interface SearchNode {
  id: string;
  name: string;
  url: string;
  stargazerCount: number;
  forkCount: number;
}

export interface SearchEdge {
  node: SearchNode;
}

export interface SearchResults {
  repositoryCount: number;
  edges: SearchEdge[];
}
