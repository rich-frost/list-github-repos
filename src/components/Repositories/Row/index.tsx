import { Tr, Td, Icon, Link } from '@chakra-ui/react';
import { GoStar, GoRepoForked } from 'react-icons/go';
import { SearchNode } from '../../../types/search';
import { BiLinkExternal } from 'react-icons/bi';

type RowProps = {
  data: SearchNode;
};

export default function Row({ data }: RowProps) {
  return (
    <Tr>
      <Td>
        {data.url ? (
          <Link href={data.url} isExternal>
            {data.name} <Icon as={BiLinkExternal} mx="2px" w={3} h={3} />
          </Link>
        ) : (
          data.name
        )}
      </Td>
      <Td>
        <Icon as={GoStar} /> {data.stargazerCount.toLocaleString()}
      </Td>
      <Td>
        <Icon as={GoRepoForked} /> {data.forkCount.toLocaleString()}
      </Td>
    </Tr>
  );
}
