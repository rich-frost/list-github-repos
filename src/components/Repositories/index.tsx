import {
  Table,
  Thead,
  Tbody,
  Tr,
  Th,
  TableContainer,
  Icon,
} from '@chakra-ui/react';
import { GoStar, GoRepoForked } from 'react-icons/go';
import { SearchEdge } from '../../types/search';
import Footer from './Footer';
import NoData from './NoData';
import Row from './Row';

type RepositoriesProps = {
  data: SearchEdge[];
};

//TODO swap out this table for react-table or similar for pagination/virtual lists etc
export default function Repositories({ data }: RepositoriesProps) {
  return (
    <TableContainer>
      <Table variant="simple">
        <Thead>
          <Tr>
            <Th>Name</Th>
            <Th>
              <Icon as={GoStar} /> Stars
            </Th>
            <Th>
              <Icon as={GoRepoForked} /> Forks
            </Th>
          </Tr>
        </Thead>
        <Tbody>
          {data.length === 0 && <NoData />}
          {!!data &&
            data.map((repo) => <Row key={repo.node.id} data={repo.node} />)}
        </Tbody>
        {!!data && data.length > 0 && <Footer />}
      </Table>
    </TableContainer>
  );
}
