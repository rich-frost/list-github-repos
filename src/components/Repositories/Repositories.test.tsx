import renderer from 'react-test-renderer';
import { SearchEdge } from '../../types/search';
import Repositories from './index';

describe('Repositories component', () => {
  const data: SearchEdge[] = [
    {
      node: {
        id: '123',
        name: 'React repo',
        stargazerCount: 100,
        forkCount: 240,
        url: 'google.com',
      },
    },
  ];

  it('renders with no data message', () => {
    const component = renderer.create(<Repositories data={[]} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders the component', () => {
    const component = renderer.create(<Repositories data={data} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
