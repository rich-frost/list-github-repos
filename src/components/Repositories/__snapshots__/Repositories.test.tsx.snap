// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repositories component renders the component 1`] = `
<div
  className="chakra-table__container css-zipzvv"
>
  <table
    className="chakra-table css-0"
  >
    <thead
      className="css-0"
    >
      <tr
        className="css-0"
      >
        <th
          className="css-0"
        >
          Name
        </th>
        <th
          className="css-0"
        >
          <svg
            className="chakra-icon css-13otjrl"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 14 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
              fillRule="evenodd"
            />
          </svg>
           Stars
        </th>
        <th
          className="css-0"
        >
          <svg
            className="chakra-icon css-13otjrl"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 10 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1a1.993 1.993 0 0 0-1 3.72V6L5 8 3 6V4.72A1.993 1.993 0 0 0 2 1a1.993 1.993 0 0 0-1 3.72V6.5l3 3v1.78A1.993 1.993 0 0 0 5 15a1.993 1.993 0 0 0 1-3.72V9.5l3-3V4.72A1.993 1.993 0 0 0 8 1zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zm3 10c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zm3-10c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"
              fillRule="evenodd"
            />
          </svg>
           Forks
        </th>
      </tr>
    </thead>
    <tbody
      className="css-0"
    >
      <tr
        className="css-0"
      >
        <td
          className="css-0"
        >
          <a
            className="chakra-link css-0"
            href="google.com"
            rel="noopener"
            target="_blank"
          >
            React repo
             
            <svg
              className="chakra-icon css-16o4ymj"
              fill="currentColor"
              focusable={false}
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m13 3 3.293 3.293-7 7 1.414 1.414 7-7L21 11V3z"
              />
              <path
                d="M19 19H5V5h7l-2-2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2h14c1.103 0 2-.897 2-2v-5l-2-2v7z"
              />
            </svg>
          </a>
        </td>
        <td
          className="css-0"
        >
          <svg
            className="chakra-icon css-13otjrl"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 14 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
              fillRule="evenodd"
            />
          </svg>
           
          100
        </td>
        <td
          className="css-0"
        >
          <svg
            className="chakra-icon css-13otjrl"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 10 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1a1.993 1.993 0 0 0-1 3.72V6L5 8 3 6V4.72A1.993 1.993 0 0 0 2 1a1.993 1.993 0 0 0-1 3.72V6.5l3 3v1.78A1.993 1.993 0 0 0 5 15a1.993 1.993 0 0 0 1-3.72V9.5l3-3V4.72A1.993 1.993 0 0 0 8 1zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zm3 10c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zm3-10c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"
              fillRule="evenodd"
            />
          </svg>
           
          240
        </td>
      </tr>
    </tbody>
    <tfoot
      className="css-0"
    >
      <tr
        className="css-0"
      >
        <td
          className="css-0"
          colSpan={3}
        >
          <div
            className="css-1bvc4cc"
          >
            <i
              className="chakra-text css-0"
            >
              Showing top 10 GitHub repos for topic, sorted by best match
            </i>
          </div>
        </td>
      </tr>
    </tfoot>
  </table>
</div>
`;

exports[`Repositories component renders with no data message 1`] = `
<div
  className="chakra-table__container css-zipzvv"
>
  <table
    className="chakra-table css-0"
  >
    <thead
      className="css-0"
    >
      <tr
        className="css-0"
      >
        <th
          className="css-0"
        >
          Name
        </th>
        <th
          className="css-0"
        >
          <svg
            className="chakra-icon css-13otjrl"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 14 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 6l-4.9-.64L7 1 4.9 5.36 0 6l3.6 3.26L2.67 14 7 11.67 11.33 14l-.93-4.74L14 6z"
              fillRule="evenodd"
            />
          </svg>
           Stars
        </th>
        <th
          className="css-0"
        >
          <svg
            className="chakra-icon css-13otjrl"
            fill="currentColor"
            focusable={false}
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              {
                "color": undefined,
              }
            }
            viewBox="0 0 10 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 1a1.993 1.993 0 0 0-1 3.72V6L5 8 3 6V4.72A1.993 1.993 0 0 0 2 1a1.993 1.993 0 0 0-1 3.72V6.5l3 3v1.78A1.993 1.993 0 0 0 5 15a1.993 1.993 0 0 0 1-3.72V9.5l3-3V4.72A1.993 1.993 0 0 0 8 1zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zm3 10c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zm3-10c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"
              fillRule="evenodd"
            />
          </svg>
           Forks
        </th>
      </tr>
    </thead>
    <tbody
      className="css-0"
    >
      <tr
        className="css-0"
      >
        <td
          className="css-0"
          colSpan={3}
        >
          <div
            className="css-gmuwbf"
          >
            <svg
              className="chakra-icon css-av9skc"
              fill="currentColor"
              focusable={false}
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                {
                  "color": undefined,
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M11.001 10h2v5h-2zM11 16h2v2h-2z"
              />
              <path
                d="M13.768 4.2C13.42 3.545 12.742 3.138 12 3.138s-1.42.407-1.768 1.063L2.894 18.064a1.986 1.986 0 0 0 .054 1.968A1.984 1.984 0 0 0 4.661 21h14.678c.708 0 1.349-.362 1.714-.968a1.989 1.989 0 0 0 .054-1.968L13.768 4.2zM4.661 19 12 5.137 19.344 19H4.661z"
              />
            </svg>
             No repositories found
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
